commit 5754713d209491301bb22d1b82cd0aedd96a3a97 (refs/changes/71/471/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2016-09-15T16:08:14+00:00 (3 years, 1 month ago)
    
    FWAPI-257 Upgrading a CN set up before June 2013 broken by FWAPI-239

diff --git a/src/fw/lib/locker.js b/src/fw/lib/locker.js
index 01450ace..afe10b22 100644
--- a/src/fw/lib/locker.js
+++ b/src/fw/lib/locker.js
@@ -32,13 +32,46 @@ var mod_lockfd = require('lockfd');
 
 // --- Globals
 
+var FW_DIR_PATH = '/var/fw';
 var FW_LOCK_FILE = '/var/fw/.lockfile';
 
+/* jsl:ignore */
+var FW_MODE = 0755;
+/* jsl:end */
+
 
 // --- Internal functions
 
+
+/**
+ * Opening the lock file should normally be fine, but in the odd event that we
+ * are trying for the first time on a system initially set up with a pre-June
+ * 2013 platform, when the /var/fw directory started shipping with the platform,
+ * and there've never been any rules on this system, then we need to take care
+ * of creating the directory.
+ */
+function tryOpen(callback) {
+    fs.open(FW_LOCK_FILE, 'w+', FW_MODE, function (err, fd) {
+        if (!err || err.code !== 'ENOENT') {
+            callback(err, fd);
+            return;
+        }
+
+        try {
+            fs.mkdirSync(FW_DIR_PATH, FW_MODE);
+            fd = fs.openSync(FW_LOCK_FILE, 'w+', FW_MODE);
+        } catch (_) {
+            callback(err);
+            return;
+        }
+
+        callback(null, fd);
+    });
+}
+
+
 function acquireLock(operation, callback) {
-    fs.open(FW_LOCK_FILE, 'w+', function (err, fd) {
+    tryOpen(function (err, fd) {
         if (err) {
             callback(err);
             return;
